### ISSUE TITLE
`unnecessary_mut_passed` gives different result depending on OS 

### ISSUE THREAD

**Issue by jaques-sam** [rust-lang/rust-clippy#10823]
### Summary

We have a debian10 (docker used: rust:1.68-buster) vs debian11 (docker used: rust:1.68-bullseye) (same on latest fedora) using opencv version v0.77.0.
cargo output says this in both cases

> Compiling opencv v0.77.0
> Compiling opencv-binding-generator v0.56.0

On the former, `delated_mask` need to be passed as `&mut` or the compiler gives the following error:

```rust
error[E0308]: mismatched types
    --> cli-tool/src/action_types/camera_output.rs:209:13
     |
208  |         imgproc::find_contours(
     |         ---------------------- arguments to this function are incorrect
209  |             &dilated_mask,
     |             ^^^^^^^^^^^^^ types differ in mutability
     |
     = note: expected mutable reference `&mut dyn ToInputOutputArray`
                        found reference `&opencv::prelude::Mat`
note: function defined here
    --> /workspaces/confero-cam-compositor/target/debug/build/opencv-a591e015a058b4d4/out/opencv/imgproc.rs:4312:9
     |
4312 |     pub fn find_contours(image: &mut dyn core::ToInputOutputArray, contours: &mut dyn core::ToOutputArray, mode: i32, method: i32, offset...
     |            ^^^^^^^^^^^^^
```

On the latter, with `&mut` we get the following clippy error:
```rust
error: the function `imgproc::find_contours` doesn't need a mutable reference
   --> cli-tool/src/action_types/camera_output.rs:209:13
    |
209 |             &mut dilated_mask,
    |             ^^^^^^^^^^^^^^^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_mut_passed
    = note: `-D clippy::unnecessary-mut-passed` implied by `-D warnings`
```

The code can be found here:
https://github.com/twistedfall/opencv-rust/blob/be022e4037cbda9491250529da1e935d173c2913/docs/imgproc.rs#LL4654C7-L4654C7

So which one is correct?

Thanks!

### Reproducer

I tried this code:

```rust
    use opencv::{core as cv, imgproc, prelude::*};
    imgproc::find_contours(
        &mut dilated_mask,
        &mut contours,
        imgproc::RetrievalModes::RETR_TREE as i32,
        imgproc::ContourApproximationModes::CHAIN_APPROX_SIMPLE as i32,
        cv::Point::new(-1, -1),
    )
    .unwrap();
```

I expected to see this happen:
  No error whatsoever

Instead, this happened:
  Either compiler error or clippy error depending on which system

### Version

```text
rustc 1.68.2 (9eb3afe9e 2023-03-27)
binary: rustc
commit-hash: 9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0
commit-date: 2023-03-27
host: x86_64-unknown-linux-gnu
release: 1.68.2
LLVM version: 15.0.6
```

**Comment by Alexendoo** [comment]
opencv 3 (debian 10) and opencv 4 (debian 11) have different signatures for that function

https://docs.opencv.org/3.4.19/d3/dc0/group__imgproc__shape.html#ga17ed9f5d79ae97bd4c7cf18403e1689a
https://docs.opencv.org/4.x/d3/dc0/group__imgproc__shape.html#gadf1ad6a0b82947fa1fe3c3d497f260e0

In the opencv 4 generated function is just an `InputArray` and so doesn't need `&mut`, but if opencv 3 needs `&mut` to compile and it works with 4 too it would be reasonable to add `#[allow(clippy::unnecessary-mut-passed)]`

**Comment by jaques-sam** [comment]
Thanks! This makes sense.On 25 May 2023, at 15:53, Alex Macleod ***@***.***> wrote:﻿
opencv 3 (debian 10) and opencv 4 (debian 11) have different signatures for that function
https://docs.opencv.org/3.4.19/d3/dc0/group__imgproc__shape.html#ga17ed9f5d79ae97bd4c7cf18403e1689a
https://docs.opencv.org/4.x/d3/dc0/group__imgproc__shape.html#gadf1ad6a0b82947fa1fe3c3d497f260e0
In the opencv 4 generated function is just an InputArray and so doesn't need &mut, but if opencv 3 needs &mut to compile and it works with 4 too it would be reasonable to add #[allow(clippy::unnecessary-mut-passed)]
